### PR TITLE
Extended terraforming to create EFS terraform resource.

### PR DESCRIPTION
**Pull request by notjames**

### CODE REVIEW

File: lib/terraforming/resource/efs.rb [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        @client = client
+      end
+
+      def tf
+        apply_template(@client, "tf/elastic_filesystem")
+      end
+
+      def tfstate
+        efsystems.inject({}) do |resources, efs|
+          attributes = {
```
Comment (dtan4) [inline lib/terraforming/resource/efs.rb]: __[rubocop]__
* Useless assignment to variable - `attributes`. [:ref](https://github.com/bbatsov/ruby-style-guide#underscore-unused-vars)

File: lib/terraforming/resource/efs.rb [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+              "creationtoken" =>  efs.creationtoken,
```
Comment (dtan4) [inline lib/terraforming/resource/efs.rb]: __[rubocop]__
* Operator `=>` should be surrounded by a single space. [:ref](https://github.com/bbatsov/ruby-style-guide#spaces-operators)

File: lib/terraforming/resource/efs.rb [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+      def initialize(client)
+        @client = client
+      end
+
+      def tf
+        apply_template(@client, "tf/elastic_filesystem")
+      end
+
+      def tfstate
+        efsystems.inject({}){|resources, efs| resources}
```
Comment (dtan4) [inline lib/terraforming/resource/efs.rb]: __[rubocop]__
* Space missing to the left of {.
* Space between { and | missing.
* Name `inject` block params `|a, e|`. [:ref](https://github.com/bbatsov/ruby-style-guide#reduce-blocks)
* Unused block argument - `efs`. If it's necessary, use `_` or `_efs` as an argument name to indicate that it won't be used. [:ref](https://github.com/bbatsov/ruby-style-guide#underscore-unused-vars)
* Space missing inside }.

File: lib/terraforming/resource/efs.rb [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+      def initialize(client)
+        @client = client
+      end
+
+      def tf
+        apply_template(@client, "tf/elastic_filesystem")
+      end
+
+      def tfstate
+        efsystems.inject({}) { |resources, _efs| resources }
```
Comment (dtan4) [inline lib/terraforming/resource/efs.rb]: __[rubocop]__
* Name `inject` block params `|a, e|`. [:ref](https://github.com/bbatsov/ruby-style-guide#reduce-blocks)

Comment (notjames) [inline lib/terraforming/resource/efs.rb]: Not really understanding this req. Checking into it more.

File: lib/terraforming/resource/efs.rb [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+      end
+
+      def tf
+        apply_template(@client, "tf/elastic_filesystem")
+      end
+
+      def tfstate
+        idx = -1
+
+        efsystems.inject({}) do  |resources, efs|
```
Comment (dtan4) [inline lib/terraforming/resource/efs.rb]: __[rubocop]__
* Unnecessary spacing detected.

File: spec/lib/terraforming/resource/efs_spec.rb [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+module Terraforming
+  module Resource
+    describe EFS do
+      let(:client) do
+        Aws::EFS::Client.new(stub_responses: true)
+      end
+
+      let(:efs_description_0) do
+          {
```
Comment (dtan4) [inline spec/lib/terraforming/resource/efs_spec.rb]: __[rubocop]__
* Use 2 (not 4) spaces for indentation. [:ref](https://github.com/bbatsov/ruby-style-guide#spaces-indentation)

File: spec/lib/terraforming/resource/efs_spec.rb [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+            creation_time: Time.parse("2016-10-24 11:42:21 -0700"),
+            creation_token: "console-0000abcd-4321-dcba-a123-d34db33f0000",
+            file_system_id: "fs-abcd1234",
+            life_cycle_state: "available",
+            name: "efs_name_1",
+            number_of_mount_targets: 3,
+            owner_id: "999999999999",
+            performance_mode: "generalPurpose",
+            size_in_bytes: {value: 23481234},
```
Comment (dtan4) [inline spec/lib/terraforming/resource/efs_spec.rb]: __[rubocop]__
* Space inside { missing. [:ref](https://github.com/bbatsov/ruby-style-guide#spaces-operators)
* Space inside } missing. [:ref](https://github.com/bbatsov/ruby-style-guide#spaces-operators)

File: spec/lib/terraforming/resource/efs_spec.rb [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+                            "tags.%" => "1",
+                            "tags.Name" => "efs_name_0"
+                        },
+                        "meta" => {},
+                        "tainted" => false
+                    },
+                    "deposed" => [],
+                    "provider" => "aws"
+                },
+                "aws_efs_file_system.efs.1" => {
```
Comment (dtan4) [inline spec/lib/terraforming/resource/efs_spec.rb]: __[rubocop]__
* Align the elements of a hash literal if they span more than one line.

File: spec/lib/terraforming/resource/efs_spec.rb [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+                            "tags.%"           => "1",
+                            "tags.Name"        => "efs_name_0"
+                        },
+                        "meta"    => {},
+                        "tainted" => false,
+                    },
+                    "deposed"  => [],
+                    "provider" => "aws",
+                },
+                "aws_efs_file_system.efs.1" => {
```
Comment (notjames) [inline spec/lib/terraforming/resource/efs_spec.rb]: This style request is not making a great deal of sense to me with respect to this data object referenced.

File: lib/terraforming/resource/efs.rb [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+      def tf
+        apply_template(@client, "tf/elastic_filesystem")
+      end
+
+      def tfstate
+        idx = -1
+
+        efsystems.inject({}) do |resources, efs|
+          idx += 1
```
Comment (dtan4) [inline lib/terraforming/resource/efs.rb]: How about this:

```ruby
efsystems.each.with_index.inject({}) do |resources, (efs, idx)| 
  # ...
  resources["aws_efs_file_system.efs.#{idx}"] = {
```

File: lib/terraforming/resource/efs.rb [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+            "provider" => "aws",
+          }
+
+          resources
+        end
+      end
+
+      private
+
+      def efsystems
```
Comment (dtan4) [inline lib/terraforming/resource/efs.rb]: `efsystems` is not a good name. This should be `file_systems`.

File: lib/terraforming/resource/efs.rb [hunk 12]
```diff
@@ -0,0 +1,62 @@
+# author: Jim Conner <snafu.x@gmail.com>
+# Nov 2016
+module Terraforming
+  module Resource
+    class EFS
```
Comment (dtan4) [inline lib/terraforming/resource/efs.rb]: Please match Ruby class name to target Terraform resource name.
In this case, class name should be `EFSFileSystem`.

## REF

https://www.terraform.io/docs/providers/aws/r/efs_file_system.html

File: lib/terraforming/resource/efs.rb [hunk 13]
```diff
@@ -0,0 +1,62 @@
+# author: Jim Conner <snafu.x@gmail.com>
+# Nov 2016
```
Comment (dtan4) [inline lib/terraforming/resource/efs.rb]: Please do not include unnecessary comments.

**Comment by dtan4** [comment]
Resource name must not contain period (`.`), so `efs.#{idx}` is invalid for resource name.

```
Errors:

  * aws_efs_file_system.efs.0: : invalid or unknown key: file_system_id
  * aws_efs_file_system.efs.0: efs.0: resource name can only contain letters, numbers, dashes, and underscores.
```

How about using other UNIQUE value, e.g. `file_system_id`?

**Comment by dtan4** [comment]
Sorry for late review 🙇

File: .gitignore [hunk 14]
```diff
@@ -6,4 +6,5 @@
 /doc/
 /pkg/
 /spec/reports/
+*.sw?
```
Comment (notjames) [inline .gitignore]: Sorry. Autocorrect sucks.

"Press not" was meant to read please note. "Fine"should have been fixed.

File: lib/terraforming/resource/efs.rb (hunk 12 above)
Comment (notjames) [inline lib/terraforming/resource/efs.rb]: Will fix

File: lib/terraforming/resource/efs.rb (hunk 13 above)
Comment (notjames) [inline lib/terraforming/resource/efs.rb]: Ok

File: lib/terraforming/resource/efs.rb (hunk 11 above)
Comment (notjames) [inline lib/terraforming/resource/efs.rb]: Ok.

**Comment by notjames** [comment]
No worries on late review. I'll fix these up and resubmit pull request.

File: .gitignore (hunk 14 above)
Comment (dtan4) [inline .gitignore]: If you're using vim, I strongly recommend to create `~/.gitignore` and write the below.

```
[._]*.s[a-w][a-z]
[._]s[a-w][a-z]
# session
Session.vim
# temporary
.netrwhist
*~
# auto-generated tag files
tags
```

_(generated by https://www.gitignore.io/api/vim)_

This gitignore affects globally in your machine. You do not have to write this settings at each repository.

Comment (notjames) [inline .gitignore]: fixed.

File: lib/terraforming/resource/efs.rb (hunk 13 above)
Comment (notjames) [inline lib/terraforming/resource/efs.rb]: fixed

File: lib/terraforming/resource/efs.rb (hunk 12 above)
Comment (notjames) [inline lib/terraforming/resource/efs.rb]: fixed

File: lib/terraforming/resource/efs.rb (hunk 10 above)
Comment (notjames) [inline lib/terraforming/resource/efs.rb]: fixed.

File: lib/terraforming/resource/efs.rb (hunk 11 above)
Comment (notjames) [inline lib/terraforming/resource/efs.rb]: fixed

Comment (notjames) [inline lib/terraforming/resource/efs.rb]: with respect to using an unique identifier, I originally used the file_system_id for the name. I was unclear on how that would work for the array in the case of multiple FSes. I'll use this advice for now and hopefully there will be no issues with it. All cards on the table: I'm still not a heavy user of terraform; or in other words, I'm just getting started with it. So, yeah.

File: lib/terraforming/resource/efs.rb [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+        file_systems.each_with_index.inject({}) do |resources, (efs,idx)|
+          attributes = {
+            "creation_token" => efs.creation_token,
+            "id" => efs.file_system_id,
+            "performance_mode" => efs.performance_mode,
+            "tags.%" => "1",
+            "tags.Name" => efs.name,
+          }
+
+          resources['%s' % [efs.file_system_id]] = {
```
Comment (dtan4) [inline lib/terraforming/resource/efs.rb]: __[rubocop]__
* Favor `format` over `String#%`. [:ref](https://github.com/bbatsov/ruby-style-guide#sprintf)

File: lib/terraforming/resource/efs.rb [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+        file_systems.each_with_index.inject({}) do |resources, (efs,idx)|
+          attributes = {
+            "creation_token" => efs.creation_token,
+            "id" => efs.file_system_id,
+            "performance_mode" => efs.performance_mode,
+            "tags.%" => "1",
+            "tags.Name" => efs.name,
+          }
+
+          resources[format('%s',efs.file_system_id)] = {
```
Comment (dtan4) [inline lib/terraforming/resource/efs.rb]: __[rubocop]__
* Space missing after comma. [:ref](https://github.com/bbatsov/ruby-style-guide#spaces-operators)

File: lib/terraforming/resource/efs.rb [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+      def initialize(client)
+        @client = client
+      end
+
+      def tf
+        apply_template(@client, "tf/elastic_filesystem")
+      end
+
+      def tfstate
+        file_systems.each_with_index.inject({}) do |resources, (efs, idx)|
```
Comment (dtan4) [inline lib/terraforming/resource/efs.rb]: __[rubocop]__
* Unused block argument - `idx`. If it's necessary, use `_` or `_idx` as an argument name to indicate that it won't be used. [:ref](https://github.com/bbatsov/ruby-style-guide#underscore-unused-vars)

File: spec/lib/terraforming/resource/efs_spec.rb [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+    }
+}
+        EOS
+        end
+      end
+
+      describe ".tfstate" do
+        it "should generate tfstate" do
+          expect(described_class.tfstate(client: client)).to eq({
+            "fs-0000abcd" => {
```
Comment (dtan4) [inline spec/lib/terraforming/resource/efs_spec.rb]: Is `aws_efs-file-system.` prefix really unnecessary? Other resources has resource name prefix; e.g. https://github.com/notjames/terraforming/blob/master/spec/lib/terraforming/resource/ec2_spec.rb#L455

File: lib/terraforming/resource/efs.rb [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+        file_systems.inject({}) do |resources, efs|
+          attributes = {
+            "creation_token" => efs.creation_token,
+            "id" => efs.file_system_id,
+            "performance_mode" => efs.performance_mode,
+            "tags.%" => "1",
+            "tags.Name" => efs.name,
+          }
+
+          resources[format('%s', efs.file_system_id)] = {
```
Comment (dtan4) [inline lib/terraforming/resource/efs.rb]: Please use string interpolation to inject a certain value in String literal.

Actually in this case, it seems that neither string interpolation or `format` is required...

Comment (notjames) [inline lib/terraforming/resource/efs.rb]: Looks like I did in fact miss the aws_efs_file_system prefix in the resource name. Also changed to interpolation (a style that I personally hate, btw, but you're the owner, so... :))

File: spec/lib/terraforming/resource/efs_spec.rb (hunk 18 above)
Comment (notjames) [inline spec/lib/terraforming/resource/efs_spec.rb]: see previous note. This is fixed now.

**Comment by dtan4** [comment]
One last thing... please rename `efs.rb` and `efs_spec.rb` to `efs_file_system.rb` and `efs_file_system_spec.rb`. Mostly in Ruby, class name and file name should be matched.

**Comment by notjames** [comment]
Will do it tonight and push the change.

**Comment by dtan4** [comment]
LGTM 👍  Thanks ❗️

**Pull request merged**